ptp/ptp4u/server/server.go

```diff
@@ -402,17 +404,27 @@ func (s *Server) findWorker(clientID ptp.PortIdentity, r *rand.Rand) *sendWorker
 	return s.sw[r.Intn(s.Config.SendWorkers)]
 }
 
-// sendGrant sends a Unicast Grant message
-func (s *Server) sendGrant(sc *SubscriptionClient, sg *ptp.Signaling, mt ptp.UnicastMsgTypeAndFlags, interval ptp.LogInterval, duration uint32) {
-	sc.UpdateGrant(sg, mt, interval, duration)
-	sc.OnceGrant()
-}
-
 // Drain traffic
 func (s *Server) Drain() {
 	if s.ctx != nil && s.ctx.Err() == nil {
 		s.cancel()
 	}
+
+	// Wait for drain to complete for up to 10 seconds
+	for i := 0; i < 100; i++ {
+		// Verifying all subscriptions are over
+		for _, w := range s.sw {
+			w.inventoryClients()
+
+			for _, subs := range w.clients {
+				if len(subs) != 0 {
+					log.Info("Still waiting for drain to finish...")
+					time.Sleep(100 * time.Millisecond)
+					break
+				}
+			}
+		}
+	}
 }
```

Comment:
oh, this is a bit ugly...

Anyway, let's maybe log how many subs we have here `log.Info("Still waiting for drain to finish...")`?

Also, as we only wait for 10s, there may be subs left - let's log how many are left, if any?

Reply:
In reality we only go once over this loop (sleep 100 mils ensures all done). But I can add additional logging of course.